lib/shopify_app/controller_concerns/token_exchange.rb

```diff
@@ -5,8 +5,18 @@ module TokenExchange
     extend ActiveSupport::Concern
     include ShopifyApp::AdminAPI::WithTokenRefetch
 
+    INVALID_SHOPIFY_ID_TOKEN_ERRORS = [
+      ShopifyAPI::Errors::CookieNotFoundError,
+      ShopifyAPI::Errors::InvalidJwtTokenError,
+    ].freeze
+
     def activate_shopify_session(&block)
-      retrieve_session_from_token_exchange if current_shopify_session.blank? || should_exchange_expired_token?
+      begin
+        retrieve_session_from_token_exchange if current_shopify_session.blank? || should_exchange_expired_token?
+      rescue *INVALID_SHOPIFY_ID_TOKEN_ERRORS => e
```

Comment:
Could we rescue these errors for the entire `activate_shopify_session` method? `with_token_refetch` might also end up raising this error, since it performs token exchange.

Reply:
I was pairing with Mike and we thought that's not worth it since we already check for validity before we call token exchange, it'd just be adding extra work for the slightest chance that it might expire within that timeframe... There are leeways in-place to ensure we don't encounter these problems..?

Reply:
Oh but what if we don't perform token exchange there because `current_shopify_session` exists and then do during `with_token_refetch`? In that case it wouldn't be double validation/rescue, right?

Reply:
hmmm I think this will cause a "Double render/redirect" error since this is around the controller action.. so if we handle the error and respond with retry or redirect to bounce page and the controller action also renders... 🤔

Reply:
If the action raises a token exchange error it will likely not have rendered yet, but you're right that it _could_ happen if the app doesn't follow the pattern of rendering/redirecting as the last thing in the action 🤔 

I looked up and saw there's a `performed?` [method](https://api.rubyonrails.org/classes/ActionController/Metal.html#method-i-performed-3F) we can call in the controller to avoid a double render/redirect. What if we bounce unless performed?